File: black-pearl-admin-bot/parser.js
const fetch = require('node-fetch');
const API = process.env.APPS_SCRIPT_URL;

const PRICE_REGEX = /^(.+?)\s+(?:по\s+)?(\d[\d\s]*)\s*(₽|руб|rub|сум)/gim;
const POSITION_REGEX = /^(\d+)\.\s+(.+?)\s+@([a-zA-Z0-9_]+)(?:\s*\/\/\s*(\d{2}\.\d{2}))?\s*$/gim;
const QUEUE_MEMBER_REGEX = /^([а-яА-Яa-zA-ZёЁ]+)\s*:\s*@?([a-zA-Z0-9_]+)?(?:\s*\/\/\s*(\d{2}\.\d{2}))?\s*$/gim;
const OLD_PRICE_REGEX = /по\s+(\d[\d\s]*)\s*(₽|руб|rub|сум)/i;
const PAY_REGEX = /@([a-zA-Z0-9_]+)\s*[-—]\s*([\d\s]+)\s*(₽|руб|rub|сум)?(?:\s*\/\/\s*(\d{2}\.\d{2}))?/g;

function detectPostType(text) {
  if (/@[\w]+\s*[-—]\s*\d+/.test(text) && !/^\d+\.\s+.+@/m.test(text)) return 'payment';
  if (/^\d+\.\s+.+?@/m.test(text)) return 'signup_positions';
  if ((/очередь/i.test(text) || /по\s+\d+\s*₽/i.test(text)) && /@[\w]+/.test(text)) return 'signup';
  return 'unknown';
}

function parsePositionsPost(text) {
  const result = {postTitle: null, hashtag: null, priceList: {}, positions: []};
  const lines = text.split('\n').map(l => l.trim());
  result.postTitle = lines.find(l => l.length > 0) || 'Без названия';
  
  const hashtagMatch = text.match(/#([a-zA-Z0-9_а-яА-Я]+)/);
  if (hashtagMatch) result.hashtag = '#' + hashtagMatch[1];
  
  const firstPositionIdx = lines.findIndex(l => /^\d+\.\s+/.test(l));
  const headerLines = lines.slice(0, firstPositionIdx > 0 ? firstPositionIdx : lines.length);
  
  headerLines.forEach(line => {
    const match = line.match(/^(.+?)\s+(?:по\s+)?(\d[\d\s]*)\s*(₽|руб|rub|сум)/i);
    if (match) {
      const name = match[1].trim().replace(/[!,]$/,'').trim();
      const price = parseInt(match[2].replace(/\s/g, ''));
      if (name && price) result.priceList[name.toLowerCase()] = {name: match[1].trim(), price};
    }
  });
  
  const blocks = splitIntoPositionBlocks(text);
  blocks.forEach(block => {
    const position = parsePositionBlock(block);
    if (position) result.positions.push(position);
  });
  
  return result;
}

function splitIntoPositionBlocks(text) {
  const blocks = [];
  const lines = text.split('\n');
  let currentBlock = null;
  
  lines.forEach(line => {
    const posMatch = line.match(/^(\d+)\.\s+(.+?)\s+@([a-zA-Z0-9_]+)(?:\s*\/\/\s*(\d{2}\.\d{2}))?\s*$/);
    if (posMatch) {
      if (currentBlock) blocks.push(currentBlock);
      currentBlock = {
        header: line, queueLines: [],
        positionNum: parseInt(posMatch[1]),
        positionName: posMatch[2].trim(),
        mainBuyer: posMatch[3],
        mainDeadline: posMatch[4] || null
      };
    } else if (currentBlock) {
      currentBlock.queueLines.push(line);
    }
  });
  
  if (currentBlock) blocks.push(currentBlock);
  return blocks;
}

function parsePositionBlock(block) {
  const position = {
    number: block.positionNum,
    name: block.positionName,
    mainBuyer: {username: block.mainBuyer, deadline: block.mainDeadline},
    queue: []
  };
  
  block.queueLines.forEach(line => {
    const clean = line.trim();
    if (!clean || /^очередь/i.test(clean)) return;
    
    const match = clean.match(/^([а-яА-Яa-zA-ZёЁ]+)\s*:\s*@?([a-zA-Z0-9_]+)?(?:\s*\/\/\s*(\d{2}\.\d{2}))?\s*$/);
    if (match) {
      const username = match[2] || null;
      if (!username) return;
      position.queue.push({member: match[1], username, deadline: match[3] || null});
    }
  });
  
  return position;
}

function parseSignupPost(text) {
  const result = {postTitle: null, price: null, entries: []};
  const lines = text.split('\n').map(l => l.trim()).filter(Boolean);
  result.postTitle = lines[0];
  
  const priceMatch = text.match(OLD_PRICE_REGEX);
  if (priceMatch) result.price = parseInt(priceMatch[1].replace(/\s/g, ''));
  
  const queues = splitByQueues(text);
  queues.forEach(q => parseQueueEntries(q.text, q.number, result.entries));
  
  return result;
}

function splitByQueues(text) {
  const queues = [];
  const lines = text.split('\n');
  let currentQueue = null;
  let currentText = [];
  
  lines.forEach(line => {
    const m = line.match(/^(\d+)\s+очередь/i);
    if (m) {
      if (currentQueue !== null && currentText.length > 0) queues.push({number: currentQueue, text: currentText.join('\n')});
      currentQueue = parseInt(m[1]);
      currentText = [];
    } else if (currentQueue !== null) {
      currentText.push(line);
    } else {
      if (currentQueue === null) currentQueue = 1;
      currentText.push(line);
    }
  });
  
  if (currentQueue !== null && currentText.length > 0) queues.push({number: currentQueue, text: currentText.join('\n')});
  return queues;
}

function parseQueueEntries(queueText, queueNumber, entries) {
  queueText.split('\n').map(l => l.trim()).filter(Boolean).forEach(line => {
    if (/^\d+\s+очередь/i.test(line)) return;
    const m = line.match(/^([А-Яа-яA-Za-z]+)\s+@([a-zA-Z0-9_]+)(?:\s*\/\/\s*(\d{2}\.\d{2}))?/);
    if (m) {
      entries.push({name: m[1], username: m[2], deadline: m[3] || null, queue: queueNumber});
      return;
    }
    const m2 = line.match(/^([А-Яа-яA-Za-z]+)\s*$/);
    if (m2) entries.push({name: m2[1], username: null, deadline: null, queue: queueNumber});
  });
}

function parsePaymentPost(text) {
  const entries = [];
  let match;
  while ((match = PAY_REGEX.exec(text)) !== null) {
    entries.push({username: match[1], amount: parseInt(match[2].replace(/\s/g, '')), deadline: match[4] || null});
  }
  return {entries};
}

async function findUserByUsername(username) {
  try {
    const res = await fetch(`${API}?action=findUser&username=${encodeURIComponent(username)}`);
    const data = await res.json();
    return data && data.telegram_id ? data.telegram_id : null;
  } catch(e) { return null; }
}

function findPriceForPosition(positionName, priceList) {
  const exact = priceList[positionName.toLowerCase()];
  if (exact) return exact.price;
  for (const key in priceList) {
    if (positionName.toLowerCase().includes(key) || key.includes(positionName.toLowerCase())) return priceList[key].price;
  }
  return null;
}

async function processPositionsPost(text, adminName) {
  const parsed = parsePositionsPost(text);
  if (!parsed.positions.length) return {success: false, error: 'Не найдено ни одной позиции'};
  
  const allEntries = [];
  parsed.positions.forEach(pos => {
    const price = findPriceForPosition(pos.name, parsed.priceList);
    allEntries.push({username: pos.mainBuyer.username, name: pos.mainBuyer.username, positionNumber: pos.number, positionName: pos.name, role: 'main', deadline: pos.mainBuyer.deadline, price});
    pos.queue.forEach(q => {
      allEntries.push({username: q.username, name: q.member, positionNumber: pos.number, positionName: pos.name, role: 'queue', deadline: q.deadline, price});
    });
  });
  
  for (const entry of allEntries) {
    if (entry.username) entry.telegramId = await findUserByUsername(entry.username);
  }
  
  const res = await fetch(API, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({action: 'createOrdersFromPositions', postTitle: parsed.postTitle, hashtag: parsed.hashtag, priceList: parsed.priceList, admin: adminName, entries: allEntries})
  });
  return res.json();
}

async function processSignupPost(text, adminName) {
  const parsed = parseSignupPost(text);
  if (!parsed.postTitle) return {success: false, error: 'Не найдено название'};
  if (!parsed.price) return {success: false, error: 'Не найдена цена (формат: "по XXX₽")'};
  if (!parsed.entries.length) return {success: false, error: 'Нет записей'};
  
  for (const entry of parsed.entries) {
    if (entry.username) entry.telegramId = await findUserByUsername(entry.username);
  }
  
  const res = await fetch(API, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({action: 'createOrdersFromPost', postTitle: parsed.postTitle, price: parsed.price, admin: adminName, entries: parsed.entries})
  });
  return res.json();
}

async function processPaymentPost(text, adminName) {
  const parsed = parsePaymentPost(text);
  if (!parsed.entries.length) return {success: false, error: 'Нет записей об оплате'};
  for (const entry of parsed.entries) entry.telegramId = await findUserByUsername(entry.username);
  
  const res = await fetch(API, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({action: 'createPaymentsFromPost', admin: adminName, entries: parsed.entries})
  });
  return res.json();
}

module.exports = {
  detectPostType, parsePositionsPost, parseSignupPost, parsePaymentPost,
  processPositionsPost, processSignupPost, processPaymentPost
};

File: black-pearl-admin-bot/parser.py
import re
import aiohttp
import os
API_URL = os.getenv('APPS_SCRIPT_URL')

# ===== ФУНКЦИИ ПАРСИНГА (без изменений) =====

def detect_post_type(text: str) -> str:
    if re.search(r'@\w+\s*[-—]\s*\d+', text) and not re.search(r'^\d+\.\s+.+@', text, re.MULTILINE):
        return 'payment'
    if re.search(r'^\d+\.\s+.+?@', text, re.MULTILINE):
        return 'signup_positions'
    if (re.search(r'очередь', text, re.IGNORECASE) or re.search(r'по\s+\d+\s*₽', text, re.IGNORECASE)) and re.search(r'@\w+', text):
        return 'signup'
    return 'unknown'

def parse_positions_post(text: str) -> dict:
    result = {'postTitle': None, 'hashtag': None, 'priceList': {}, 'positions': []}
    lines = [l.strip() for l in text.split('\n')]
    
    result['postTitle'] = next((l for l in lines if l), 'Без названия')
    
    hashtag_match = re.search(r'#([a-zA-Z0-9_а-яА-Я]+)', text)
    if hashtag_match:
        result['hashtag'] = '#' + hashtag_match.group(1)
    
    first_pos_idx = next((i for i, l in enumerate(lines) if re.match(r'^\d+\.\s+', l)), len(lines))
    for line in lines[:first_pos_idx]:
        match = re.match(r'^(.+?)\s+(?:по\s+)?(\d[\d\s]*)\s*(?:₽|руб|rub|сум)', line, re.IGNORECASE)
        if match:
            name = match.group(1).strip().rstrip('!,').strip()
            price = int(match.group(2).replace(' ', ''))
            if name and price:
                result['priceList'][name.lower()] = {'name': match.group(1).strip(), 'price': price}
    
    blocks = []
    current_block = None
    for line in lines:
        pos_match = re.match(r'^(\d+)\.\s+(.+?)\s+@([a-zA-Z0-9_]+)(?:\s*\/\/\s*(\d{2}\.\d{2}))?\s*$', line)
        if pos_match:
            if current_block: blocks.append(current_block)
            current_block = {
                'positionNum': int(pos_match.group(1)),
                'positionName': pos_match.group(2).strip(),
                'mainBuyer': pos_match.group(3),
                'mainDeadline': pos_match.group(4),
                'queueLines': []
            }
        elif current_block:
            current_block['queueLines'].append(line)
    if current_block: blocks.append(current_block)
    
    for block in blocks:
        position = {
            'number': block['positionNum'],
            'name': block['positionName'],
            'mainBuyer': {'username': block['mainBuyer'], 'deadline': block['mainDeadline']},
            'queue': []
        }
        for line in block['queueLines']:
            clean = line.strip()
            if not clean or re.match(r'^очередь', clean, re.IGNORECASE): continue
            m = re.match(r'^([а-яА-Яa-zA-ZёЁ]+)\s*:\s*@?([a-zA-Z0-9_]+)?(?:\s*\/\/\s*(\d{2}\.\d{2}))?\s*$', clean)
            if m and m.group(2):
                position['queue'].append({'member': m.group(1), 'username': m.group(2), 'deadline': m.group(3)})
        result['positions'].append(position)
        
    return result

def parse_signup_post(text: str) -> dict:
    """
    Парсит пост записи с очередями.
    Цена — первое число, за которым идёт знак ₽.
    """
    result = {'postTitle': None, 'price': None, 'entries': []}
    lines = [l.strip() for l in text.split('\n') if l.strip()]
    
    if not lines:
        return result
    
    # Заголовок — первая непустая строка
    result['postTitle'] = lines[0]
    
    # Цена: ищем число, за которым идёт ₽ (с пробелом или без)
    # Паттерн: число (возможно с пробелами-разделителями тысяч) + ₽
    price_match = re.search(r'(\d[\d\s]*?)\s*₽', text)
    if price_match:
        price_str = price_match.group(1).replace(' ', '')
        if price_str:
            result['price'] = int(price_str)
    
    # Разбиваем на очереди
    queues = []
    current_queue = None
    current_lines = []
    
    for line in lines:
        queue_match = re.match(r'^(\d+)\s+очередь', line, re.IGNORECASE)
        if queue_match:
            if current_queue is not None and current_lines:
                queues.append({'number': current_queue, 'lines': current_lines})
            current_queue = int(queue_match.group(1))
            current_lines = []
        elif current_queue is not None:
            current_lines.append(line)
        else:
            if current_queue is None:
                current_queue = 1
                current_lines.append(line)
    
    if current_queue is not None and current_lines:
        queues.append({'number': current_queue, 'lines': current_lines})
    
    # Парсим записи в каждой очереди
    for queue in queues:
        for line in queue['lines']:
            # Паттерн 1: "Имя @username //дата"
            m = re.match(
                r'^([А-Яа-яA-Za-zёЁ]+)\s+@([a-zA-Z0-9_]+)(?:\s*\/\/\s*(\d{2}\.\d{2}))?\s*$', 
                line
            )
            if m:
                result['entries'].append({
                    'name': m.group(1),
                    'username': m.group(2),
                    'deadline': m.group(3),
                    'queue': queue['number'],
                    'telegramId': None
                })
                continue
            
            # Паттерн 2: свободный слот (без username) — пропускаем 
            # TODO Админы тоже должны писать свои ники или потом подставлять их из базы
            m2 = re.match(r'^([А-Яа-яA-Za-zёЁ]+)\s*(?:[🥰💖❤️🔥✨🌸]|\s)*$', line)
            if m2:
                print(f"⏭️ Занято админом в очереди {queue['number']}: {m2.group(1)}")
                continue
    
    return result

def parse_payment_post(text: str) -> dict:
    entries = []
    for m in re.finditer(r'@([a-zA-Z0-9_]+)\s*[-—]\s*([\d\s]+)\s*(?:₽|руб|rub|сум)?(?:\s*\/\/\s*(\d{2}\.\d{2}))?', text):
        entries.append({
            'username': m.group(1),
            'amount': int(m.group(2).replace(' ', '')),
            'deadline': m.group(3),
            'telegramId': None
        })
    return {'entries': entries}

def find_price_for_position(position_name: str, price_list: dict) -> int:
    exact = price_list.get(position_name.lower())
    if exact: return exact['price']
    for key, val in price_list.items():
        if position_name.lower() in key or key in position_name.lower():
            return val['price']
    return 0

# ===== НОВАЯ ФУНКЦИЯ: АВТОДОБАВЛЕНИЕ ПОЛЬЗОВАТЕЛЕЙ =====
async def ensure_users_in_db(usernames: list, session):
    """
    Проверяет пользователей и создаёт отсутствующих.
    Возвращает (user_map, new_users) — словарь и список новых.
    """
    user_map = {}
    new_users = []
    
    for username in usernames:
        if not username:
            continue
        
        try:
            async with session.post(API_URL, json={
                'action': 'upsertUserByUsername',
                'username': username
            }) as resp:
                result = await resp.json()
                
                if result.get('success'):
                    user_map[username] = result.get('telegram_id')
                    if result.get('action') == 'created':
                        new_users.append(username)
                        print(f"✅ Создан пользователь: @{username}")
                    else:
                        print(f"🔍 Найден: @{username}")
                else:
                    print(f"❌ Ошибка для @{username}: {result.get('error')}")
                    user_map[username] = None
                    
        except Exception as e:
            print(f"❌ Ошибка запроса для @{username}: {e}")
            user_map[username] = None
    
    return user_map, new_users
